gmso/formats/mcf.py

```diff
@@ -0,0 +1,705 @@
+                    dihedral_style,
+                    0.5
+                    * dihedral.connection_type.parameters["k0"]
+                    .in_units("kJ/mol")
+                    .value,
+                    0.5
+                    * dihedral.connection_type.parameters["k1"]
+                    .in_units("kJ/mol")
+                    .value,
+                    0.5
+                    * dihedral.connection_type.parameters["k2"]
+                    .in_units("kJ/mol")
+                    .value,
+                    0.5
+                    * dihedral.connection_type.parameters["k3"]
+                    .in_units("kJ/mol")
+                    .value,
+                )
+            )
+        elif dihedral_style == "charmm":
+            mcf.write(
+                "{:s}  "
+                "{:10.5f}  "
+                "{:10.5f}  "
+                "{:10.5f}\n".format(
+                    dihedral_style,
+                    dihedral.connection_type.parameters["k"]
+                    .in_units("kJ/mol")
+                    .value,
+                    dihedral.connection_type.parameters["n"],
+                    dihedral.connection_type.parameters["phi_eq"]
+                    .in_units(u.degrees)
+                    .value,
+                )
+            )
+        elif dihedral_style == "harmonic":
+            mcf.write(
+                "{:s}  "
+                "{:10.5f}  "
+                "{:10.5f}\n".format(
+                    dihedral_style,
+                    0.5
+                    * dihedral.connection_type.parameters["k"]
+                    .in_units("kJ/mol")
+                    .value,
+                    dihedral.connection_type.parameters["phi_eq"]
+                    .in_units(u.degrees)
+                    .value,
+                )
+            )
+
+        else:
+            raise GMSOError(
+                "Unsupported dihedral style for Cassandra MCF writer"
+            )
+
+
+def _write_improper_information(mcf, top):
+    """Write the impropers in the system.
+
+    Parameters
+    ----------
+    mcf : file object
+        The file object of the Cassandra mcf being written
+    top : Topology
+        Topology object
+    """
+
+    header = (
+        "\n!Improper Format\n"
+        "!index i j k l type parameters\n"
+        '!type="harmonic", parms=force_constant equilibrium_improper\n'
+        "\n# Improper_Info\n"
+    )
+
+    mcf.write(header)
+    mcf.write("{:d}\n".format(len(top.impropers)))
+
+    improper_type = "harmonic"
+    for i, improper in enumerate(top.impropers):
+        mcf.write(
+            "{:<4d}  {:<4d}  {:<4d}  {:<4d}  {:<4d}"
+            "  {:s}  {:8.3f}  {:8.3f}\n".format(
+                i + 1,
+                improper.atom1.idx + 1,
+                improper.atom2.idx + 1,
+                improper.atom3.idx + 1,
+                improper.atom4.idx + 1,
+                improper_type,
+                improper.type.psi_k * KCAL_TO_KJ,
```

Comment:
I feel like i must be overlooking it, but where is this conversion factor stored, and could we use unyt here?

Reply:
Nope, you're not overlooking anything. I have the improper section commented out for now, so that function isn't used. I'll update it when I come back and add 1-4 scaling and add the tests for bonded systems. I'd vote to get this merged for now and then as we continue to expand the tests for GMSO I'll come back and finish off these few details. You'll notice there are some "TODO:"s scattered throughout the code.